lib/assert.js

```diff
@@ -931,6 +933,87 @@ assert.doesNotMatch = function doesNotMatch(string, regexp, message) {
   internalMatch(string, regexp, message, doesNotMatch);
 };
 
+const noop = () => {};
+
+class Tracker {
+
+  constructor() {
+    this.callChecks = new SafeSet();
+  }
+
+  calls(fn, exact = 1) {
+    if (process._exiting)
+      throw new ERR_CALLS_UNAVAILABLE_DURING_EXIT();
+    if (typeof fn === 'number') {
+      exact = fn;
+      fn = noop;
+    } else if (fn === undefined) {
+      fn = noop;
+    }
+
+    validateUint32(exact, 'exact', true);
+
+    const context = {
+      exact,
+      actual: 0,
+      // eslint-disable-next-line no-restricted-syntax
+      stackTrace: new Error(),
+      name: fn.name || 'calls'
+    };
+    this.callChecks.add(context);
+    const callChecks = this.callChecks;
+
+    return function() {
+      // Increment call count
+      context.actual++;
+      if (context.actual === context.exact) {
+        // Once function has reached its call count remove it from
+        // callChecks set to prevent memory leaks
+        callChecks.delete(context);
+      }
+      if (context.actual > context.exact) {
+        // If call count exceeds exact throw an error
+        const message = `Expected the ${context.name} function to be ` +
+        `executed ${context.exact} time(s) but was executed ` +
+        `${context.actual} time(s).`;
+
+        throw new AssertionError({
+          actual: context.actual,
+          expected: context.exact,
+          message,
+          operator: context.name,
+        });
+      }
+      return fn.apply(this, arguments);
+    };
+  }
+
+  verify() {
+    for (const context of this.callChecks) {
+      // const console = require('console');
+      // console.log(context);
+      if (context.actual !== context.exact) {
+        const message = `Expected the ${context.name} function to be ` +
+        `executed ${context.exact} time(s) but was executed ` +
+        `${context.actual} time(s).`;
+        const console = require('console');
```

Comment:
I don't think this should log anything.

Reply:
I have it print the stack trace to the actual function that wasn't called, I thought it would be better than having a stack trace of the verify function. If that's not needed I'll change that part

Reply:
I don't think there is anything wrong with storing that information and giving it back to the user somehow, but printing things to the console without permission generally annoys users.

Reply:
Ah ok fair enough, cheers